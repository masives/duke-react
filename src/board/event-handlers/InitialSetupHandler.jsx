// @flow
/* eslint-disable no-param-reassign */
import configuration from '../../config/index.json';

class InitialSetupHandler {
  mutableBoard: BoardCells;

  initialSetupState = {
    white: {
      dukeDrawn: false,
      knightsDrawn: 0
    },
    black: {
      dukeDrawn: false,
      knightsDrawn: 0
    }
  };

  handleInitialSetup = (cell: CellState, color: string) => {
    let result;
    if (!this.initialSetupState[color].dukeDrawn) {
      console.log('coordinates', cell.coordinates);
      if (cell.coordinates.x !== this.initialDukeRow(color)) {
        console.log('wrong cell');
        result = false;
      }
      if (cell.coordinates.x === this.initialDukeRow(color)) {
        this.mutableBoard[cell.coordinates.x][cell.coordinates.y].color = color;
        this.mutableBoard[cell.coordinates.x][cell.coordinates.y].unitType = 'duke';
        this.initialSetupState[color].dukeDrawn = true;
        console.log('board', this.mutableBoard);
        result = { boardCells: this.mutableBoard, currentPlayer: 'black' };
      }
    }
    return result;
  };

  initialDukeRow = (color: string) => (color === 'white' ? 0 : configuration.boardSize.height - 1);

  generateCells = () => {
    const Cells: BoardCells = new Array(configuration.boardSize.width);
    for (let i = 0; i < configuration.boardSize.width; i += 1) {
      Cells[i] = new Array(configuration.boardSize.width);
      for (let j = 0; j < configuration.boardSize.height; j += 1) {
        Cells[i][j] = {
          coordinates: {
            x: i,
            y: j
          },
          color: '',
          unitType: '',
          polarity: '',
          selected: false,
          targeted: false
        };
      }
    }
    this.mutableBoard = JSON.parse(JSON.stringify(Cells));
    return Cells;
  };
}

export default InitialSetupHandler;
