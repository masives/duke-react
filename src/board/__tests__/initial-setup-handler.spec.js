// @flow

import InitialSetupHandler from '../initial-setup-handler';
import emptyBoard from './test-helpers/board/board-empty.json';
import emptyDukeCoordinates from './test-helpers/duke-coordinates/empty.json';
import PLAYER_COLOR from '../enums/player-color';

test('Should initialize correctly', () => {
  const initialSetupHandler = new InitialSetupHandler();

  expect(initialSetupHandler.initialSetupState).toEqual(
    expect.objectContaining({
      white: expect.objectContaining({
        dukeDrawn: false,
        footmanDrawn: 0
      }),
      black: expect.objectContaining({
        dukeDrawn: false,
        footmanDrawn: 0
      })
    })
  );
});

test('Should return error message when trying to set in wrong row', () => {
  const initialSetupHandler = new InitialSetupHandler();
  const invalidTargetCell: CellState = {
    coordinates: {
      row: 1,
      col: 2
    },
    color: '',
    unitType: '',
    startingSide: true,
    state: ''
  };
  const result = initialSetupHandler.handleInitialSetup(
    invalidTargetCell,
    PLAYER_COLOR.WHITE,
    emptyBoard,
    emptyDukeCoordinates
  ); // one argument is missing here

  expect(result.message).toBe('wrong row - please choose the top row');
});

test("Should return board with duke and update it's coordinates[WHITE]", () => {
  // given
  const initialSetupHandler = new InitialSetupHandler();
  const currentPlayer = PLAYER_COLOR.WHITE;
  const validWhiteDukeTargetCell: CellState = {
    coordinates: {
      row: 0,
      col: 2
    },
    color: '',
    unitType: '',
    startingSide: true,
    state: ''
  };

  // when
  const result = initialSetupHandler.handleInitialSetup(
    validWhiteDukeTargetCell,
    currentPlayer,
    emptyBoard,
    emptyDukeCoordinates
  );

  // then
  expect(result.board[validWhiteDukeTargetCell.coordinates.row][validWhiteDukeTargetCell.coordinates.col]).toEqual(
    expect.objectContaining({
      color: currentPlayer,
      unitType: 'duke'
    })
  );
  expect(result.dukeCoordinates).toEqual(
    expect.objectContaining({
      white: expect.objectContaining({
        row: validWhiteDukeTargetCell.coordinates.row,
        col: validWhiteDukeTargetCell.coordinates.col
      })
    })
  );
});

test("Should return board with duke and update it's coordinates[BLACK]", () => {
  // given
  const initialSetupHandler = new InitialSetupHandler();
  initialSetupHandler.initialSetupState = {
    white: {
      dukeDrawn: true,
      footmanDrawn: 2
    },
    black: {
      dukeDrawn: false,
      footmanDrawn: 0
    }
  };
  const currentPlayer = PLAYER_COLOR.BLACK;
  const validBlackDukeTargetCell: CellState = {
    coordinates: {
      row: 5,
      col: 2
    },
    color: '',
    unitType: '',
    startingSide: true,
    state: ''
  };

  // when
  const result = initialSetupHandler.handleInitialSetup(
    validBlackDukeTargetCell,
    currentPlayer,
    emptyBoard,
    emptyDukeCoordinates
  );

  // then
  expect(result.board[validBlackDukeTargetCell.coordinates.row][validBlackDukeTargetCell.coordinates.col]).toEqual(
    expect.objectContaining({
      color: currentPlayer,
      unitType: 'duke'
    })
  );
  expect(result.dukeCoordinates).toEqual(
    expect.objectContaining({
      black: expect.objectContaining({
        row: validBlackDukeTargetCell.coordinates.row,
        col: validBlackDukeTargetCell.coordinates.col
      })
    })
  );
});
