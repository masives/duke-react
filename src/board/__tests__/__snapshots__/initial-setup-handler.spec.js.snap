// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return board with duke and update it's coordinates[WHITE] 1`] = `
Array [
  Array [
    Object {
      "color": "",
      "coordinates": Object {
        "col": 0,
        "row": 0,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 1,
        "row": 0,
      },
      "startingSide": true,
      "state": "TARGETED_DRAW",
      "unitType": "",
    },
    Object {
      "color": "white",
      "coordinates": Object {
        "col": 2,
        "row": 0,
      },
      "startingSide": true,
      "state": "",
      "unitType": "duke",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 3,
        "row": 0,
      },
      "startingSide": true,
      "state": "TARGETED_DRAW",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 4,
        "row": 0,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 5,
        "row": 0,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
  ],
  Array [
    Object {
      "color": "",
      "coordinates": Object {
        "col": 0,
        "row": 1,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 1,
        "row": 1,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 2,
        "row": 1,
      },
      "startingSide": true,
      "state": "TARGETED_DRAW",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 3,
        "row": 1,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 4,
        "row": 1,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 5,
        "row": 1,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
  ],
  Array [
    Object {
      "color": "",
      "coordinates": Object {
        "col": 0,
        "row": 2,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 1,
        "row": 2,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 2,
        "row": 2,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 3,
        "row": 2,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 4,
        "row": 2,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 5,
        "row": 2,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
  ],
  Array [
    Object {
      "color": "",
      "coordinates": Object {
        "col": 0,
        "row": 3,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 1,
        "row": 3,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 2,
        "row": 3,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 3,
        "row": 3,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 4,
        "row": 3,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 5,
        "row": 3,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
  ],
  Array [
    Object {
      "color": "",
      "coordinates": Object {
        "col": 0,
        "row": 4,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 1,
        "row": 4,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 2,
        "row": 4,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 3,
        "row": 4,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 4,
        "row": 4,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 5,
        "row": 4,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
  ],
  Array [
    Object {
      "color": "",
      "coordinates": Object {
        "col": 0,
        "row": 5,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 1,
        "row": 5,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 2,
        "row": 5,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 3,
        "row": 5,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 4,
        "row": 5,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
    Object {
      "color": "",
      "coordinates": Object {
        "col": 5,
        "row": 5,
      },
      "startingSide": true,
      "state": "",
      "unitType": "",
    },
  ],
]
`;
